clientv3/balancer.go

```diff
@@ -140,48 +138,47 @@ func (b *simpleBalancer) Up(addr grpc.Address) func(error) {
 		return func(err error) {}
 	}
 
-	if len(b.upEps) == 0 {
+	if b.pinAddr == "" {
 		// notify waiting Get()s and pin first connected address
 		close(b.upc)
 		b.pinAddr = addr.Addr
+		// notify client that a connection is up
+		b.readyOnce.Do(func() { close(b.readyc) })
+		// close opened connections that are not pinAddr
+		// this ensures only one connection is open per client
+		b.notifyCh <- []grpc.Address{addr}
 	}
-	b.upEps[addr.Addr] = struct{}{}
-
-	// notify client that a connection is up
-	b.readyOnce.Do(func() { close(b.readyc) })
 
 	return func(err error) {
 		b.mu.Lock()
-		delete(b.upEps, addr.Addr)
-		if len(b.upEps) == 0 && b.pinAddr != "" {
+		if b.pinAddr == addr.Addr {
 			b.upc = make(chan struct{})
-		} else if b.pinAddr == addr.Addr {
-			// choose new random up endpoint
-			for k := range b.upEps {
-				b.pinAddr = k
-				break
-			}
+			b.pinAddr = ""
+			b.notifyCh <- b.addrs
 		}
 		b.mu.Unlock()
 	}
 }
 
 func (b *simpleBalancer) Get(ctx context.Context, opts grpc.BalancerGetOptions) (grpc.Address, func(), error) {
-	var addr string
+	var (
+		addr   string
+		closed bool
+	)
+	b.mu.RLock()
```

Comment:
I looked through the grpc code. This is racey:
1. upc is not closed; no pinned address
2. client issues an rpc, calling invoke(), which calls Get(), enters for loop, blocks
3. clientconn.Close() calls balancer.Close(); closed = true
4. for loop never exits since ctx is the context passed in by the client and may not be canceled (I was mistaken thinking it was controlled by grpc)

So it seems you had it right the first time with:
```go
select {
case <-upc:
default: close(upc)
}
```
in `Close`, but it also needs a `closed` check in the for loop here so it can exit.